### PR TITLE
Adding Type2 targeting package and tests

### PR DESCRIPTION
**Pull request by swernli**
This changes continues building on the work in feature/decomp to enable targeting alternate gate sets via packages specifying the supported set of intrinsics and decompositions for that gate set. This introduces Type2, which targets a set of intrinsics that includes `SWAP` and the 3 Pauli Ising gates. It also adds support for executing against this target package in the QDK simulator, and shares some of the existing intrinsic unitary tests to validate behavior of the package.

See #249 for the rundown on what work is part of this overall feature.

### CODE REVIEW

**Comment by swernli** [comment]
One thing these changes do not yet support is the `EntryPointDriver`, which for now pulls in the QsharpCore-based simulator as a compilation dependency. I think there is a way to refactor it such that it takes a generic dependency on `SimulatorBase` and then uses which ever simulator comes in the target package the executable was compiled against, but I'm not sure how to make that work with the Toffoli simulator and Resource Estimator as default options (those are not present in Type2 yet). This shouldn't affect the current PR, but definitely something we need to figure out as part of surfacing targeting packages as a finished feature.
@samarsha for FYI.

**Comment by swernli** [comment]
To make this large and unwieldy PR a little bit easier to review, I'll call out the files that a new or significantly modified here...

- [src/Simulation/Common/Simulators.Type2.Dev.props](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-f513eb71de42c407d46dc6fa0e38e2b9)

- [src/Simulation/Simulators.Core/Microsoft.Quantum.Simulators.csproj](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-25faec016091dade212364d3fe09fa2e)

- [src/Simulation/Simulators.Tests/OperationsTestHelper.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-498a3c68479c71a5c79dc326ff8a05f9)

- [src/Simulation/Simulators.Tests/OperationsTestHelperSimSupport.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-4b623e9112f7de62b3fc957e294be896)

- [src/Simulation/Simulators.Tests/TracerHelper.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-1754395f7d9edce7bc4bc9c4321b76d4)

- src/Simulation/Simulators.Type2.Tests/*

- src/Simulation/Simulators.Type2/*

- [src/Simulation/Simulators/QuantumSimulator/Checks.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-3264b3989590877f1868f483cae0ec0c)

- [src/Simulation/Simulators/QuantumSimulator/IsingXX.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-6f1b3897d22a4670f14fbdaf65f279b8)

- [src/Simulation/Simulators/QuantumSimulator/IsingYY.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-68df6da8fbd4d95ffaeb054429931bb1)

- [src/Simulation/Simulators/QuantumSimulator/IsingZZ.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-38b35ceac36d11ad5737ddcefaf930fb)

- [src/Simulation/Simulators/QuantumSimulator/Rx.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-376bd69756ff160246eab52ba4550027)

- [src/Simulation/Simulators/QuantumSimulator/Ry.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-ed0888400138bbca1167d2a40ed39075)

- [src/Simulation/Simulators/QuantumSimulator/Rz.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-10c4984b3f77b98c2ef0bd14e33e7ee7)

- [src/Simulation/Simulators/QuantumSimulator/SWAP.cs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-fc41841dbd0b9fe2e0545f04ca74eee2)

- [src/Simulation/TargetDefinitions/Decompositions/ExpFracFromIsing.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-ce241ed191edc6a15b7d66ad311644f7)

- [src/Simulation/TargetDefinitions/Decompositions/ExpFracFromIsing.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-0ff5ee2bab5d5d8c1c45e1110aae73e0)

- [src/Simulation/TargetDefinitions/Decompositions/MResetXWithNoReuse.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-18efe55969222c44d12e7b5ee89ba6d5)

- [src/Simulation/TargetDefinitions/Decompositions/MResetYWithNoReuse.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-f30d8bebd3c126c8ca1db10ee0a4ff49)

- [src/Simulation/TargetDefinitions/Decompositions/MResetYWithNoReuse.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-5643d660f9616f184c593c12522cec72)

- [src/Simulation/TargetDefinitions/Decompositions/MeasureWithNoReuse.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-eead33921506b75189dc1200dcd61bfc)

- [src/Simulation/TargetDefinitions/Decompositions/R.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-298f37156d04264ae4e835ee63f4e39c)

- [src/Simulation/TargetDefinitions/Decompositions/ResetWithNoReuse.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-65a6d6939bdfe0b2b3db175accd96f0d)

- [src/Simulation/TargetDefinitions/Decompositions/Utils.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-f20ccbfd920a5f25a7709577fb2e1bc2)

- [src/Simultors/TargetDefinitions/Intrinsic/Checks.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-bb6bb46b9aa8dd33d4ed2f205ca90274)

- [src/Simultors/TargetDefinitions/Intrinsic/IsingXX.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-66e9d08dc3e4ca6249c6ff3984d3882a)

- [src/Simultors/TargetDefinitions/Intrinsic/IsingYY.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-987bf4cb6e629e0b0f10dea58c0ebe75)

- [src/Simultors/TargetDefinitions/Intrinsic/IsingZZ.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-bd3f88a5193f21b00544acf252713523)

- [src/Simultors/TargetDefinitions/Intrinsic/M.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-50171a5221df7449b331d4cddf58a656)

- [src/Simultors/TargetDefinitions/Intrinsic/Rx.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-41676f2cd87712dacef90cbb4cf2b06a)

- [src/Simultors/TargetDefinitions/Intrinsic/Ry.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-8a944ac6a1a12400dc10fa7638e12901)

- [src/Simultors/TargetDefinitions/Intrinsic/Rz.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-241b8d450886e2073ff0166cb956c35e)

- [src/Simultors/TargetDefinitions/Intrinsic/SWAP.qs](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-25f3665db5319b37eab2d5758756e57a)

- [src/Simulation/TargetDefinitions/TargetPackages/Type2.Package.props](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-233678157037e3e3be828fb427f29d49)

- [src/Simulation/Type2Core/Microsoft.Quantum.Type2.Core.csproj](https://github.com/microsoft/qsharp-runtime/pull/363/files#diff-d3190b40c4f260130d2cabf9ff495ade)

**Comment by samarsha** [comment]
> One thing these changes do not yet support is the `EntryPointDriver`, which for now pulls in the QsharpCore-based simulator as a compilation dependency. I think there is a way to refactor it such that it takes a generic dependency on `SimulatorBase` and then uses which ever simulator comes in the target package the executable was compiled against, but I'm not sure how to make that work with the Toffoli simulator and Resource Estimator as default options (those are not present in Type2 yet). This shouldn't affect the current PR, but definitely something we need to figure out as part of surfacing targeting packages as a finished feature.

Does this mean EntryPointDriver will need to create the simulator via reflection in the same way that it creates the target quantum machine for Azure Quantum?

**Comment by swernli** [comment]
> Does this mean EntryPointDriver will need to create the simulator via reflection in the same way that it creates the target quantum machine for Azure Quantum?

That’s certainly one possibility, yeah. The other idea I had but haven’t tried out yet is reworking the dependency so that the entry point code only depends on `SimulatorBase` and then it just depends on the common package. Then each simulator package would pull in the entry point driver as a dependency, instead of the other way around.

**Comment by samarsha** [comment]
> Then each simulator package would pull in the entry point driver as a dependency, instead of the other way around.

That seems strange, since it's possible to use the simulators without needing the command-line interface that the EntryPointDriver provides? However, another option is that the Q# executable project would depend on both a simulator package and the driver package, and give simulator factories to the driver at runtime.

How many different simulators will there be? My understanding was that we would still have just one simulator for all execution targets, is that incorrect?

**Comment by swernli** [comment]
> That seems strange, since it's possible to use the simulators without needing the command-line interface that the EntryPointDriver provides? However, another option is that the Q# executable project would depend on both a simulator package and the driver package, and give simulator factories to the driver at runtime.

That's a cool idea. I was trying to think about how something like that might work, so I would really appreciate any ideas or advice you have on the design there.

> How many different simulators will there be? My understanding was that we would still have just one simulator for all execution targets, is that incorrect?

Unfortunately, I'm disappointed by how this part of the work turned out. I was hoping to have a single simulator that would support all targets, and I still think we should keep an eye out for the opportunities to refactor everything into one simulator to handle all targeting packages. I'll give a summary here of the issues I ran into with supporting multiple targeting packages on the same simulator...

Each targeting package collects the subset of intrinsics and decompositions that comprise it. The simulator needs to implement the intrinsics in C# so that they can wrap the appropriate calls into the native C++ simulator. However, some operations that are intrinsic for one package are decompositions for another, and in that case if the compiled simulator included C# implementations of those operations they would end up overriding the corresponding decomposition. To pick a specific example, `Exp` is an intrinsic operation for the QsharpCore package and so has an implementation in the simulator that inherits from the Q# declaration, `public class QSimExp : Quantum.Intrinsic.Exp`. This works because the body intrinsic `Exp` is just an abstract base class that gets overridden by the simulator implementation. However, `Exp` in the Type2 package is supposed to be a decomposition on top of the Ising gates, so it is fully implemented in Q# instead of abstract. But if that package is built along with the core simulator, the implementation of `QSimExp` that defines itself as a child of the `Exp` class ends up overriding the decomposition with the wrapper for the native simulator. Technically, the circuit will be run correctly, but the decomposition does not get used, so anything like qubit counts or gate counts won't really be correct, and tracing the execution will be confusing. If the goal of a simulator is just to calculate the overall result of the quantum algorithm by any means then this doesn't really matter. But if the goal is to actually simulate the sequence of operations defined by the Q# then it should include the defined decompositions and not take any short cuts. I tried to find a way around this, but as long as the design for intrinsic implementation uses inheritance and overrides I don't think we can avoid this problem. The only path forward I could come up with is ensuring that the simulator for a given target package only includes implementations for that package's intrinsics and nothing else, which necessitates a simulator per package. At least by utilizing the same patterns I used for the targeting packages we can avoid duplicating the implementation, but there will still have to be separate projects for each simulator to match the separate target packages.

**Comment by swernli** [comment]
@samarsha Rather than figure the entry point handling all as part of this PR, I've added it as an additional ToDo on the #249 so that I can ensure it gets addressed before this feature branch is merged into master. I can try a couple of the ideas we've had so far and get that together as a separate PR soon.

**Comment by samarsha** [comment]
> That's a cool idea. I was trying to think about how something like that might work, so I would really appreciate any ideas or advice you have on the design there.

Sure, I haven't thought that much about it but I could elaborate if that'd be helpful, maybe it would be easier to chat more about it offline.

> Unfortunately, I'm disappointed by how this part of the work turned out. I was hoping to have a single simulator that would support all targets, and I still think we should keep an eye out for the opportunities to refactor everything into one simulator to handle all targeting packages. I'll give a summary here of the issues I ran into with supporting multiple targeting packages on the same simulator...

Thanks for the explanation. If I understood correctly, does that mean that if the simulator could tell if a particular operation had a Q# implementation or was an intrinsic, and could decide whether to implement something based on that information, then there could be only one simulator for all targets? (Or equivalently, if a Q# implementation of an operation had higher priority over the simulator implementation?)

**Comment by swernli** [comment]
> If I understood correctly, does that mean that if the simulator could tell if a particular operation had a Q# implementation or was an intrinsic, and could decide whether to implement something based on that information, then there could be only one simulator for all targets? (Or equivalently, if a Q# implementation of an operation had higher priority over the simulator implementation?)

Yes, that is exactly what we would need. Inheritance with abstract methods and overrides doesn't give us a way to do that. The only mechanism C# offers that prevents a given implementation of a method from being overridden is `sealed` but rather than suppressing the child class override it would result in a compile time failure, which also doesn't work for this usage. I did spend some time trying to think of a schema whereby the generated classes are actually children of the simulator classes instead of the other way around such that the generated implementations always take precedence, but I didn't come up with a reasonable way to make such a hierarchy work.

**Comment by alan-geller** [comment]
> > If I understood correctly, does that mean that if the simulator could tell if a particular operation had a Q# implementation or was an intrinsic, and could decide whether to implement something based on that information, then there could be only one simulator for all targets? (Or equivalently, if a Q# implementation of an operation had higher priority over the simulator implementation?)
> 
> Yes, that is exactly what we would need. Inheritance with abstract methods and overrides doesn't give us a way to do that. The only mechanism C# offers that prevents a given implementation of a method from being overridden is `sealed` but rather than suppressing the child class override it would result in a compile time failure, which also doesn't work for this usage. I did spend some time trying to think of a schema whereby the generated classes are actually children of the simulator classes instead of the other way around such that the generated implementations always take precedence, but I didn't come up with a reasonable way to make such a hierarchy work.

I think you can override the current mechanism by modifying the `InitBuiltinOperations` method on `SimulatorBase`, or alternatively by changing the simulator class initialization and registering the Q# decompositions as overrides (using `AbstractFactory.Register`).

**Comment by swernli** [comment]
> I think you can override the current mechanism by modifying the `InitBuiltinOperations` method on `SimulatorBase`, or alternatively by changing the simulator class initialization and registering the Q# decompositions as overrides (using `AbstractFactory.Register`).
Oh, that's an interesting idea! I think I could try enlightening `InitBuiltinOperations` to only register overrides of base types that are abstract, and then the simulator could be a single implementation. Thanks @alan-geller, I'm going to try that now!

**Comment by swernli** [comment]
@alan-geller Ok, I tried the approach you suggested and still didn't quite work, but it shows promise. The issue I ran into is that the simulator still defines it's classes as inheriting from the intrinsic operations, as in `class QSimExp : Quantum.Intrinsic.Exp`. This means that the simulator project needs to be compiled with references to a targeting package that defines those classes. But once it's compiled with one (either QsharpCore or Type2Core) it cannot later be pulled in as a dependency with the other. When I try to do that, in the tests for example, it will complain that it can't resolve the conflict between the identically named classes in the two dlls. I was thinking I could try to disconnect the simulator from the core package by having the intrinsic implementations inherit directly from `Unitary` and `ICallable`, but then that would require reworking the way the intrinsics are discovered and registered in the simulator initialization and it would require the simulator classes define all the class methods instead of just the abstract ones, all while loosening the type safety we have right now. I'm likely going to adapt this change and push it up anyway, but for now it will still require two simulator projects instead of just one, which doesn't really fix the problem we were hoping to address.

**Comment by swernli** [comment]
Turns out dotnet does not support multiple .csproj files in the same folder, so my attempts at a clever combination of the two simulator folders was only accidentally working. Thankfully it failed in the `Build mac` pipeline to draw my attention to that; the mac dotnet must more aggressively parallelize the builds, preventing that approach from "working" as consistently.

**Comment by swernli** [comment]
@alan-geller @samarsha I think I've gotten these changes to the local maximum for the time being. Please let me know if you've got any other feedback or ideas for me to follow up on. I'm hoping I can get this merged today and move to the next target package. By way of summary, here's the issues we've identified as out of scope for this PR that I hope to handle later:

- EntryPointDriver should not have a compile-time dependency on the simulator, as that would require a new driver per target package/simulator package. I added this to the task list in #249 so it gets addressed as part of finishing this feature.

- One simulator should be able to act as implementation for many target packages. This is a challenging design problem, as the current use of inheritance means we must compile a simulator per target package. We have some ideas about how this might be address, but so far all of them imply some redesign and heavy reliance on some new stub generation (hopefully helped by Alan's new tooling).

**Review by alan-geller — APPROVED** [review]
A few comments, nothing huge. I really like the overall PR, I think it really cleans up a lot of our infrastructure a lot.

**Comment by swernli** [comment]
> Why not Adj + Ctl? Exp is Adj+Ctl, so this could be, unless there's a reason for it not to be. 

That's a good catch, thanks! The Ising gates are sort of special cases of a two qubit Exp rotation, so should be adjointable with the negative angle. I'll make sure to add that in!

File: src/Simulation/Simulators/QCTraceSimulator/Circuits/Primitive.IsingXX.qs
Comment (swernli) [reply src/Simulation/Simulators/QCTraceSimulator/Circuits/Primitive.IsingXX.qs:5]: I've fixed these to all be `Adj` as well. Thanks!

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (cgranade) [inline src/Simulation/TargetDefinitions/Decompositions/Utils.qs:135]: This looks like Microsoft.Quantum.Arrays.Subarray from the standard library; should that be pulled back into runtime as well?

File: src/Simulation/Simulators/QuantumSimulator/SWAP.cs
Comment (swernli) [reply src/Simulation/Simulators/QuantumSimulator/SWAP.cs:50]: Agreed, and it actually does happen in Q# when executing against the QsharpCore (see src\simulation\TargetDefinitions\Decompostions\SWAP.qs) and this code won't be used. This file is only used when executing Q# compiled with the Type2Core against our simulator, where SWAP isn't native. It's still a little odd, and more complete fix would be update our native simulator to support the union of intrinsics across all target packages, but for now it's easier to fake it in this wrapper.

File: src/Simulation/TargetDefinitions/Decompositions/MResetXWithNoReuse.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/MResetXWithNoReuse.qs:26]: For each of these "WithNoReuse" variants, the platform in question doesn't support reusing qubits after measurement, so a reset after measurement is not necessary. Further, calling `Reset` might trigger an error because it might internally use the qubit, which would count as a use of after measurement. 

I kept the comments the same to match the existing documentation of the method, but I think it's an interesting point to ask: should the doc comments on target packages always match the default or shift depending on the underlying target behavior?

File: src/Simulation/TargetDefinitions/Decompositions/MeasureWithNoReuse.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/MeasureWithNoReuse.qs:46]: I added the length check, and a comment explaining why we don't need to undo putting the qubit into the Z basis, since this is for a platform where qubits cannot be reused after measurement.

File: src/Simulation/TargetDefinitions/Decompositions/MResetXWithNoReuse.qs
Comment (alan-geller) [reply src/Simulation/TargetDefinitions/Decompositions/MResetXWithNoReuse.qs:26]: I might add a non-doc comment just to explain to someone reading the code why the code and doc comments don't match.

I think having different doc comments for different implementations of the same class would break the docs system :-).

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (alan-geller) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:135]: Good question! I'll defer to @cgranade .

File: src/Simulation/TargetDefinitions/Decompositions/R.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/R.qs:31]: Good point! This code is adapted from some earlier research code, and I didn't notice that refactoring opportunity. I agree that a single if-elif-else statement would work better!

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:40]: Agreed! I noticed this on a later pass in a different branch, but I'll go ahead and incorporate that change here.

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:135]: Oh, that's a good idea! I don't think this change is the place to do it, but I do agree that as we find ourselves reimplementing useful helper functions from the library in the core or foundation packages we should seriously consider pulling the corresponding library function in.

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:144]: Agreed, I can combine this into a single generic helper, which per suggestion from @cgranade I can model on `Subarray` for now, and replace with a call to that when/if the function gets moved into qsharp-runtime somewhere.

File: src/Simulation/TargetDefinitions/Decompositions/Utils.qs
Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:120]: Oh my, I think you are correct. It worries me that didn't cause any failures... Let me make that change and rerun the tests to confirm.

Comment (swernli) [reply src/Simulation/TargetDefinitions/Decompositions/Utils.qs:120]: Actually, that change makes the tests fail, so we must not understand the convention of this helper. I'll put it back.

**Comment by swernli** [comment]
Thanks for all the feedback and reviews!

**Pull request merged**
